packages/graphiql/src/components/QueryEditor.js

```diff
@@ -53,28 +53,33 @@ export class QueryEditor extends React.Component {
     this.cachedValue = props.value || '';
   }
 
-  componentDidMount() {
-    // Lazily require to ensure requiring GraphiQL outside of a Browser context
-    // does not produce an error.
-    const CodeMirror = require('codemirror');
-    require('codemirror/addon/hint/show-hint');
-    require('codemirror/addon/comment/comment');
-    require('codemirror/addon/edit/matchbrackets');
-    require('codemirror/addon/edit/closebrackets');
-    require('codemirror/addon/fold/foldgutter');
-    require('codemirror/addon/fold/brace-fold');
-    require('codemirror/addon/search/search');
-    require('codemirror/addon/search/searchcursor');
-    require('codemirror/addon/search/jump-to-line');
-    require('codemirror/addon/dialog/dialog');
-    require('codemirror/addon/lint/lint');
-    require('codemirror/keymap/sublime');
-    require('codemirror-graphql/hint');
-    require('codemirror-graphql/lint');
-    require('codemirror-graphql/info');
-    require('codemirror-graphql/jump');
-    require('codemirror-graphql/mode');
-
+  async componentDidMount() {
+    const { default: CodeMirror } = await import(
+      /* webpackChunkName: "codemirror" */
+      /* webpackMode: "lazy" */
+      /* webpackPrefetch: true */
+      /* webpackPreload: true */
+      'codemirror'
+    );
+    await Promise.all([
+      import('codemirror/addon/hint/show-hint'),
+      import('codemirror/addon/comment/comment'),
+      import('codemirror/addon/edit/matchbrackets'),
+      import('codemirror/addon/edit/closebrackets'),
+      import('codemirror/addon/fold/foldgutter'),
+      import('codemirror/addon/fold/brace-fold'),
+      import('codemirror/addon/search/search'),
+      import('codemirror/addon/search/searchcursor'),
+      import('codemirror/addon/search/jump-to-line'),
+      import('codemirror/addon/dialog/dialog'),
+      import('codemirror/addon/lint/lint'),
+      import('codemirror/keymap/sublime'),
+      import('codemirror-graphql/esm/hint'),
+      import('codemirror-graphql/esm/lint'),
+      import('codemirror-graphql/esm/info'),
+      import('codemirror-graphql/esm/jump'),
+      import('codemirror-graphql/esm/mode'),
+    ]);
```

Comment:
which is problematic and why?

Reply:
Because the `componentDidUpdate` method may be called and exit before the `componentDidMount` method completes, which makes the flow of the code very unintuitive and opens us up to potential race conditions (if not now, then in the future). See #994 for a better explanation/fix.